feat: Add Support for Qwen3-vl and Qwen2.5-vl Video Mode

vincentkoc opened the pull request:
<img width="200" height="200" alt="image" src="https://github.com/user-attachments/assets/6d2a4ae1-4898-4266-818b-8ae58731a772" />
<p>

**Adds comprehensive support for video understanding capabilities to Qwen3-VL and Qwen2.5-VL multimodal models in ollama**. This enables the models to analyze video content by extracting frames with temporal awareness and processing them through the vision encoder with proper temporal patch grouping. (as per [Qwens official implementation](https://github.com/QwenLM/Qwen3-VL)). This PR also resolves https://github.com/ollama/ollama/issues/12926.

Changes:
- Added Documentation Updates
- Added new `video.go` in `models/imageproc` to support video mode with supported models
- Model specific embedding and tokenizer support for video and temporal embeddings `qwen3-vl` and `qwen2.5-vl`
- API spec updated to support video in many formats base64, video url, and image frames (with temporal information)
- Token usage calculated properly with video tokens
- UI now supports video upload on supported models

Technical Details:
- Qwen3-VL
  - Uses Interleaved-MRoPE (Multidimensional Rotary Position Embeddings) with explicit temporal dimensions
  - Position embeddings: [time, height, width, extra] in 4D to avoid GGML's 5D tensor limitation
  - Better suited for long videos, precise timestamp localization, and complex temporal reasoning
- Qwen2.5-VL
  - Uses standard RoPE with spatial-only position embeddings
  - Temporal information encoded through patch embeddings (3D convolutions)
  - Faster processing, simpler architecture for general video understanding

API Usage Examples:

```bash
# Generate API with base64 video
curl http://localhost:11434/api/generate -d '{
  "model": "qwen3-vl",
  "prompt": "What is happening in this video?",
  "videos": ["<base64-encoded-video>"]
}

# Chat API with video
curl http://localhost:11434/api/chat -d '{
  "model": "qwen2.5-vl",
  "messages": [{
    "role": "user",
    "content": "Describe the actions in this video",
    "videos": ["<base64-encoded-video>"]
  }]
}

# OpenAI-compatible endpoint with video URL
curl http://localhost:11434/v1/chat/completions -d '{
  "model": "qwen3-vl",
  "messages": [{
    "role": "user",
    "content": [
      {"type": "text", "text": "What is in this video?"},
      {"type": "video_url", "video_url": {"url": "https://example.com/video.mp4"}}
    ]
  }]
}'```

vincentkoc committed:
Update types.go

vincentkoc committed:
Update qwen3vl.go

vincentkoc committed:
Update openai.go

vincentkoc committed:
feat: image processor for video with temporal sequencing

vincentkoc committed:
Merge branch 'main' into feat/video-support

vincentkoc committed:
Update runner.go

vincentkoc committed:
Update imageprocessor.go

vincentkoc committed:
Update model.go

vincentkoc committed:
Update model_vision.go

vincentkoc committed:
Update prompt.go

vincentkoc committed:
Update openai.go

vincentkoc committed:
Update qwen3vl.go

vincentkoc committed:
Create video.go

vincentkoc committed:
Update imageprocessor.go

vincentkoc committed:
Update model_vision.go

vincentkoc committed:
Update model.go

vincentkoc committed:
test: add ffmpeg

vincentkoc committed:
Create video_test.go

vincentkoc committed:
chore: docs for video

vincentkoc committed:
Update ui.go

vincentkoc committed:
Update openai.go

vincentkoc committed:
Update imageprocessor.go

vincentkoc committed:
chore: lint

vincentkoc committed:
Update openai.go

vincentkoc committed:
Update video-support.md

vincentkoc committed:
feat: video for qwen25vl

vincentkoc committed:
chore: lint

vincentkoc committed:
chore: lint

vincentkoc committed:
Update video.go

vincentkoc committed:
Update video_test.go

vincentkoc committed:
chore: changes to UI/app for video files

vincentkoc commented:
@jmorganca once i rebuild the UI and verify will send a demo example of UI plus endpoints running locally! Will attach the loom to this PR and ping you.

vincentkoc committed:
docs: update based on current implementation

vincentkoc committed:
Update api.md

vincentkoc commented:
@jmorganca also raised https://github.com/ollama/ollama/pull/12965 for the filepicker UI issue on MacOS

vincentkoc committed:
Merge branch 'main' into feat/video-support

NickM-27 commented:
The PR mentions it but I don't see any examples of how to submit images with temporal information?

NickM-27 commented:
Thanks, so that seems to suggest that `sample_fps` is the key to tell it what fps the frames were taken from. Would that be passed in as an option to Ollama API?

I agree it is less standard, definitely helpful though when you have a stronger way of knowing which frames from a video are most important to the understanding of it.

jmorganca commented:
@vincentkoc sorry for the delay, and thanks for the gentle ping. One worry I have is bringing in ffmpeg as a dependency, as it must be installed prior to this functionality working. Do you know if there are alternate ways we could extract frames – perhaps a pure Go library for this?

jessegross commented:
Having a runtime dependency on ffmpeg that requires the user to install it is going to make this pretty hard to merge. Ollama's goal is for things to just work, so this introduces too much potential for surprises. In general, dependencies are not ideal but it might be OK if everything is self contained in the install of Ollama.

vincentkoc commented:
> Having a runtime dependency on ffmpeg that requires the user to install it is going to make this pretty hard to merge. Ollama's goal is for things to just work, so this introduces too much potential for surprises. In general, dependencies are not ideal but it might be OK if everything is self contained in the install of Ollama.

@jessegross more than happy to help, let me know if you have any ideas of lightweight librarys/packages that could be added to the compiled version to avoid needed binaries

vincentkoc committed:
Update openai.go

vincentkoc committed:
Update model_vision.go

vincentkoc committed:
Merge branch 'main' into feat/video-support

vincentkoc commented:
@yejingx i patched the conflict to reuse the four slices before temporal/video handling, should be good now. Also found a bug in to API route with remote loading could lead to DDOS so i set a cap.

yejingx commented:
> @yejingx i patched the conflict to reuse the four slices before temporal/video handling, should be good now. Also found a bug in to API route with remote loading could lead to DDOS so i set a cap.

I tried it and it works fine. Thanks.

finaljrad-IR reviewed (approved):
automation

jessegross commented:
As previously mentioned, requiring the user to install ffmpeg is a blocker to merging this.

vincentkoc commented:
> As previously mentioned, requiring the user to install ffmpeg is a blocker to merging this.

@jessegross sorry not sure if you saw my comment? I'm happy to rework the solution in some way, just dont want to butcher the PR in a direction that dose not fit ollama.

> @jessegross more than happy to help, let me know if you have any ideas of lightweight librarys/packages that could be added to the compiled version to avoid needed binaries

I'm happy to package in build-time or use an alternative approach. Would love to merge this as a huge part of the community keeps messaging me since getting this PR up.

vincentkoc committed:
chore: added support for thrid-party ffmpeg build

vincentkoc committed:
chore: added support for thrid-party ffmpeg build (github actions)

vincentkoc committed:
chore: added support for thrid-party ffmpeg build (docker)

vincentkoc committed:
chore: added support for thrid-party ffmpeg build (go-astiav)

vincentkoc committed:
refactor: go-astiav wired in for video support

vincentkoc committed:
chore: added support for thrid-party ffmpeg build (ignores)

vincentkoc committed:
Merge branch 'main' into feat/video-support

vincentkoc committed:
fix: add support for video input on CLI

vincentkoc committed:
fix: finalized builds for ffmpeg lib based build

vincentkoc committed:
Update test.yaml

vincentkoc committed:
fix: builds working on weindows

vincentkoc committed:
Update test.yaml

vincentkoc committed:
chore: fix ci/cd for video

vincentkoc committed:
Update test.yaml

vincentkoc commented:
@jmorganca ready for testing and re-review 👍

vincentkoc commented:
> I tried to build the container image with 
> 
> ```sh
> 
> podman build -t ollama:rocm --build-arg=FLAVOR=rocm --build-arg=CGO_CFLAGS='-O3' --build-arg=CGO_CXXFLAGS='-O3' .
> 
> ```
> 
> and got this error
> 
> ```
> 
> [13/18] STEP 14/14: RUN --mount=type=cache,target=/root/.cache/go-build     go build -trimpath -buildmode=pie -tags "ffmpeg,cgo" -o /bin/ollama .
> 
> # github.com/asticode/go-astiav
> 
> /root/go/pkg/mod/github.com/asticode/go-astiav@v0.39.0/audio_fifo.go:3:10: fatal error: libavutil/audio_fifo.h: No such file or directory
> 
>     3 | //#include <libavutil/audio_fifo.h>
> 
>       |          ^~~~~~~~~~~~~~~~~~~~~~~~
> 
> compilation terminated.
> 
> ```

@tibequadorian look at the build scripts in my PR. You need to compile the required libraries which podman is reporting as missing. Best to use the in-built ollama build scripts which have been updated to take this into account.

vincentkoc committed:
refactor: merge conflicts

vincentkoc committed:
Merge branch 'main' into feat/video-support

vincentkoc commented:
@tibequadorian ill test on podman and resolve build issues and the merge conflicts and report back :)

vincentkoc committed:
Update Dockerfile

vincentkoc committed:
Merge branch 'main' into feat/video-support

vincentkoc committed:
Merge branch 'main' into feat/video-support

vincentkoc commented:
any updates @jmorganca @jessegross @finaljrad-IR ? there was another new merge conflict which i resolved, worried this PR is going be stale. addressed all the concerns raised and there is much request from community for support, happy to add other models etc as additional prs following.

let me know if any other concerns/questions.

vincentkoc committed:
Merge branch 'main' into feat/video-support

vincentkoc committed:
Fix openai test image constant name mismatch

hs4jkg9-alt commented:
Does it work for QW3.5 and Gemma4?

vincentkoc commented:
I'm going to refactor and also add Gemma4 support ;)

vincentkoc committed:
Merge branch 'main' into feat/video-support

vincentkoc committed:
cmd: fix multimodal merge regressions

vincentkoc committed:
test: tolerate windows ccache install failures

vincentkoc committed:
app/ui: align media validation with upload flow

lchunleo commented:
Is this feature out ? Thanks

vincentkoc commented:
> Is this feature out ? Thanks

Nope, seems like Ollama team will not merge ;(

hs4jkg9-alt commented:
So why not merge this? That's a very essential feature. 
Infuriating.

LeonKraim commented:
Is there any reason for no merge?
